impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/AbstractSourcedGraphService.java

```diff
@@ -104,4 +234,90 @@ protected void removeRelationship(Constants.Type typeInSource, Relationships.Wel
     protected abstract String getProposedId(Blueprint b);
 
     protected abstract Filter[] initNewEntity(Vertex newEntity, Blueprint blueprint);
+
+    private static class BFSTransitiveClosureSpliterator implements Spliterator<Vertex> {
+        private final Direction direction;
+        private final String[] edgeLabels;
+        Deque<Iterator<Vertex>> pending;
+
+        private BFSTransitiveClosureSpliterator(Iterator<Vertex> pending, Direction direction,
+                                                String[] edgeLabels) {
+            this.direction = direction;
+            this.edgeLabels = edgeLabels;
+            this.pending = new ArrayDeque<>(Arrays.asList(pending));
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super Vertex> action) {
+            Iterator<Vertex> queueToProcess = peekNonEmpty();
+            if (queueToProcess == null) {
+                return false;
+            }
+
+            Vertex toProcess = queueToProcess.next();
+
+            pending.add(toProcess.getVertices(direction, edgeLabels).iterator());
+
+            action.accept(toProcess);
```

Comment:
Nice! I am not sure if it'll work also with Titan, though. I think the vertices are lazily fetched when needed. It works with our current impl, because it's just a net of objects.

Have you considered using some Gremlin-fu for the transitive closure?

```Groovy
g = new TinkerGraph()
v1 = g.addVertex([name: 'one'])
v2 = g.addVertex([name: 'two'])
v3 = g.addVertex([name: 'three'])
v4 = g.addVertex([name: 'four'])
v5 = g.addVertex([name: 'five'])

v1.addEdge('contains', v2)
v2.addEdge('contains', v3)
v3.addEdge('contains', v4)
v4.addEdge('contains', v5)

v1.out('contains').loop(1){true}{true}.name
==>two
==>three
==>four
==>five
```